spec/hotel_spec.rb

```diff
@@ -0,0 +1,371 @@
+      expect(@hotel.see_available_rooms_by_date(Date.new(2019, 3, 11), Date.new(2019, 3, 14)).last.number).must_equal 20
+      expect(@hotel.see_available_rooms_by_date(Date.new(2019, 3, 11), Date.new(2019, 3, 14))[0].number).must_equal 4
+    end
+  end
+
+  describe "reserve_room method" do
+    let (:hotel) { HotelSystem::Hotel.new }
+    let (:room) { hotel.all_rooms[0] }
+
+    it "can add the reservation to the @all_reservations array" do
+      hotel.reserve_room(
+        room,
+        Date.new(2019, 3, 11),
+        Date.new(2019, 3, 14),
+        "Sam"
+      )
+      hotel.reserve_room(
+        room,
+        Date.new(2019, 3, 11),
+        Date.new(2019, 3, 14),
+        "Sam"
+      )
+      hotel.reserve_room(
+        room,
+        Date.new(2019, 3, 11),
+        Date.new(2019, 3, 14),
+        "Sam"
+      )
+
+      expect(hotel.all_reservations.length).must_equal 3
+      expect(hotel.all_reservations[0]).must_be_kind_of HotelSystem::Reservation
+      expect(hotel.all_reservations[1]).must_be_kind_of HotelSystem::Reservation
+      expect(hotel.all_reservations[2]).must_be_kind_of HotelSystem::Reservation
+    end
+
+    it "can add the reservation to the assigned room's reservation array" do
+      hotel.reserve_room(
+        room,
+        Date.new(2019, 3, 11),
+        Date.new(2019, 3, 14),
+        "Sam"
+      )
+      expect(room.reservations.length).must_equal 1
+      expect(room.reservations[0]).must_be_kind_of HotelSystem::Reservation
+    end
+
+    it "can assign a room, that doesn't have other conflicting reservations, to the reservation if the selected room is not available" do
+      hotel.reserve_room(
+        room,
+        Date.new(2019, 3, 10),
+        Date.new(2019, 3, 13),
+        "Jon"
+      )
+      hotel.reserve_room(
+        room,
+        Date.new(2019, 3, 11),
+        Date.new(2019, 3, 14),
+        "Sam"
+      )
+      hotel.reserve_room(
+        room,
+        Date.new(2019, 3, 11),
+        Date.new(2019, 3, 14),
+        "Dee"
+      )
+
+      expect(hotel.all_reservations[0].room.number).must_equal 1
+      expect(hotel.all_reservations[1].room.number).must_equal 2
+      expect(hotel.all_reservations[2].room.number).must_equal 3
+      expect(hotel.all_reservations.length).must_equal 3
+    end
+
+    it "raises an error if no rooms are available at all during specified dates" do
+      20.times do
+        hotel.reserve_room(
+          room,
+          Date.new(2019, 3, 10),
+          Date.new(2019, 3, 13),
+          "Jon"
+        )
+      end
+
+      expect do
+        hotel.reserve_room(
+          room,
+          Date.new(2019, 3, 11),
+          Date.new(2019, 3, 14),
+          "Zoe"
+        )
+      end.must_raise ArgumentError
+    end
+  end
+
+  describe "room_available? method" do
+    before do
+      @hotel = HotelSystem::Hotel.new
+      @hotel.all_rooms.slice!(2..19)
```

Comment:
This is a clever way to make this test work. Another option would be to pass in the number of rooms to the `Hotel` constructor, possibly with a default value of 20.